Tour-Collective-Devs/tc_backend
# Rm refactor employer model to auth

## Link to Ticket

Autolink format: hashtag and number, ex: #1094

Closes #9 

## Description of Proposed Changes
refactored employer model into an authenticated user with login/logout/registration endpoints

## Steps to Test

Outline the steps to test

```sh
git fetch --all
git checkout rm-refactor-employer-model-to-auth
```

1. pull down code
1. re run `pip install -r requirements.txt` many new python packages have been installed to handle user authentication
1. nuke migrations and database. this step is very important, major database changes are in this pr
1. remake migrations you will need to make migrations for multiple apps. run them in this order
 ```
... makemigrations api
... makemigrations employers
... makemigrations
... migrate
```
1. look at new code in new app named 'employers' this apps sole responsibility is to handle creation and handling of the employers endpoint. The model there inherits from an authenticated model.
1. run django server and make sure no errors have happened
1. navigate to .../api/employer/registration and make a new employer
1. go into database and confirm that employers password is being stored as a hash
1. navigate to .../api/employer-auth/login and type in the user info you just created
1. you should receive a response of a user token that can be stored in the frontend client app
1. at this point create a superuser using `python manage.py createsuperuser`
1. login at .../admin and view that the login token you received is being stored in the server
1. navigate to .../api/employer-auth/logout and click the 'post' button at the bottom of the form
1. you should see a message confirming the logout
1. go back into the admin site and check to see the user token has been removed. 


## Impacted Areas in Application
basically everything. Old employer files in the api app were removed and the heavy lifting of creating the employer model moved to the new employers app. 

List general components of the application that this PR will affect:
api
employers
admin

## Mentions @username

Tag users that need to review this code

## Testing

[ ] There are new unit tests in this PR, and I verify that there is full coverage of all new code.
[ ] I certify that all existing tests pass

## Deploy Notes

Notes regarding deployment the contained body of work.  These should note any
db migrations, etc.

## Definition of Done

1. The project must be fully documented. This includes the following:
1. Complete README that documents the steps to install the code, how to install any dependencies, or system configuration needed.
1. Every class must be documented with purpose, author, and methods.
1. Every method must be documented with purpose and argument list - which itself must contain a short purpose for each argument.
1. The project must be able to run fully, and without errors, on each teammate's system.
1. Fulfills every requirement.
1. Every line of code has been peer reviewed.
1. For projects that require unit testing, core functionality must be identified and have at least one test for each
